experiment/scheduler.py

```diff
@@ -121,6 +140,436 @@ def end_expired_trials(experiment_config: dict):
+    return bool(get_pending_trials(experiment).first())
+
+
+def any_running_trials(experiment):
+    """Returns True if there are any running trials in |experiment|."""
+    return bool(get_running_trials(experiment).first())
+
+
+class TrialInstanceManager:  # pylint: disable=too-many-instance-attributes
+    """Manager for trial instances.
+    Public methods of this are safe to call in preemptible and nonpreemptible
+    experiments alike though the main purpose of this class is to manage
+    preempted trials.
+    This class object should be created at the start of scheduling and the
+    handle_preempted_trials method should be called in the scheduling loop.
+    See the docstring for handle_preempted_trials for how it works.
+    """
+    # Hard limit on the number of nonpreemptibles we will use. This bounds
+    # costs.
+    MAX_NONPREEMPTIBLES = 500
+
+    # The maximum fraction of total trials in the experiment that can be done
+    # using preemptibles. This helps bound the cost in unexpected situations.
+    NONPREEMPTIBLES_FRACTION = 1 / 20
+
+    MAX_FRACTION_FOR_NONPREEMPTIBLES = 1 / 4
+
+    # How many times the number of trials we need preemptibles can we launch.
+    MAX_PREEMPTIBLES_MULTIPLIER = 2
+
+    # How long can we keep trying preemptibles before we have to switch to a
+    # nonpreemptibles or stopping the experiment.
+    PREEMPTIBLE_WINDOW_MULTIPLIER = 3
+
+    def __init__(self, num_trials, experiment_config):
+        self.experiment_config = experiment_config
+        self.num_trials = num_trials
+
+        # Bound for the number of nonpreemptibles we can start if the experiment
+        # specified preemptible_runners.
+        self.max_nonpreemptibles = min(
+            math.ceil(self.num_trials * self.NONPREEMPTIBLES_FRACTION),
+            self.MAX_NONPREEMPTIBLES)
+        logger.info('Max nonpreemptibles: %d.', self.max_nonpreemptibles)
+
+        # Bound for the number of preemptibles we can start if the experiment
+        # specified preemptible_runners.
+        self.max_preemptibles = (self.num_trials *
+                                 self.MAX_PREEMPTIBLES_MULTIPLIER)
+        logger.info('Max nonpreemptibles: %d.', self.max_preemptibles)
+
+        # Attributes for preemptible retry window. The preemptible retry window
+        # is a time period that starts when the last initial trial is started.
+        # It determines how long we can retry preempted trials using
+        # preemptibles. This bounds the length of time an experiment lasts.
+        self.preemptible_window = (experiment_config['max_total_time'] *
+                                   self.PREEMPTIBLE_WINDOW_MULTIPLIER)
+        self._initial_trials = list(
+            get_experiment_trials(experiment_config['experiment']))
+        self._max_time_started = None
+
+        self.preempted_trials = {}
+        self.preemptible_starts_futile = False
+
+        self.base_resource_url = (
```

Comment:
I renamed this base_target_link and I still keep it as an attribute to avoid having to format the same string every time I call get_instance_from_preempted_operation. But I moved the formatting to a function in gce.py and moved get_instance_from_preempted_operation to there as well.